### ISSUE TITLE
Redirect status code history?

### ISSUE THREAD

**Issue by mallardduck** [guzzle/guzzle#1710]
_I've searched and didn't seem to find anything on this, if I'm missing something let me know._

I noticed that #1183 added a feature for redirect history, but it only tracks the URL locations not the HTTP status codes. I'm wondering if it would be possible to add history for status codes as well? Would a PR for a solution be accepted if submitted?

I was thinking of a few ways to try and pull it off, but the easiest seems like adding it on to the current RedirectMiddleware provided with Guzzle. It could just be added an additional header similar to the current one being added; maybe something like:

```
    const STATUS_HISTORY_HEADER = 'X-Guzzle-Redirect-Status-History';


    private function withTracking(PromiseInterface $promise, $uri)
    {
        return $promise->then(
            function (ResponseInterface $response) use ($uri) {
                // Note that we are pushing to the front of the list as this
                // would be an earlier response than what is currently present
                // in the history header.
                $historyHeader = $response->getHeader(self::HISTORY_HEADER);
                $statusHeader = $response->getHeader(self::STATUS_HISTORY_HEADER);
                array_unshift($historyHeader, $uri);
                array_unshift($statusHeader, $response->getStatusCode());
                return $response->withHeader(self::HISTORY_HEADER, $historyHeader)->withHeader(self::STATUS_HISTORY_HEADER, $statusHeader);
            }
        );
    }

```

**Comment by mallardduck** [comment]
Update: Actually it looks like this method wouldn't end up getting proper status codes. I've tested it out and it seems like you end up with all 200's.

**Comment by sagikazarmark** [comment]
That sounds a little bit odd. All those status codes should be 3xx.

**Comment by mallardduck** [comment]
I think I figured out my issue; I seem to have it working now that I'm passing the status into the withTracking methoid like the uri. I'll submit a PR soon!

**Comment by sagikazarmark** [comment]
What's the point in preserving HTTP Status codes? What's the added information (besides it's a redirection)?

**Comment by mallardduck** [comment]
I see the main benefit being that by tracking the status codes we can use Guzzle in a similar way to using Curl. Using `curl -IL` can get me useful information on how a request gets from point A to point B. Redirect paths are not always as simple as A->B, when it gets complicated and looks like A->C->E->D->B then you benefit from having status codes.

Personally I think that knowing the Location header is only half the story of a redirect, knowing the status code can be pretty useful in a handful of situations. My current use case calls for needing both the locations and codes provided. By having the specific code we can weigh the severity of a redirect for [SEO purpose](https://seo-hacker.com/301-302-redirect-affect-seo/). 

TL;DR: 
Tracking the status codes provides useful SEO and debugging information. Using a 302 in place of a 301 on accident can be pretty bad; by tracking the codes Guzzles redirect tracking can be more helpful! 🚀 

Hope that helps provide some background for why this might be a benefit. Again if there's not enough demand or benefit to add this to core I can probably just implement this as a local Middleware instead of adding it to this one.

**Comment by sagikazarmark** [comment]
Thanks for the detailed response. Let's continue the discussion in the PR.

**Comment by mallardduck** [comment]
I'm so happy I was able to work with you to get this merged in. Now that it's in master I will set this as resolved since the feature will be included in the next version!
